pennylane/hf/dipole.py

```diff
@@ -0,0 +1,310 @@
+        dy = anp.einsum("qr,rs,st->qt", coeffs.T, moment_matrix(mol.basis_set, 1, 1)(*args), coeffs)
+        dz = anp.einsum("qr,rs,st->qt", coeffs.T, moment_matrix(mol.basis_set, 1, 2)(*args), coeffs)
+
+        core_x, core_y, core_z = anp.array([0]), anp.array([0]), anp.array([0])
+
+        if core is None and active is None:
+            return (core_x, core_y, core_z), (dx, dy, dz)
+
+        for i in core:
+            core_x = core_x + 2 * dx[i][i]
+            core_y = core_y + 2 * dy[i][i]
+            core_z = core_z + 2 * dz[i][i]
+
+        dx = dx[anp.ix_(active, active)]
+        dy = dy[anp.ix_(active, active)]
+        dz = dz[anp.ix_(active, active)]
+
+        return (core_x, core_y, core_z), (dx, dy, dz)
+
+    return _dipole_integrals
+
+
+def fermionic_dipole(mol, cutoff=1.0e-12, core=None, active=None):
+    r"""Return a function that builds the fermionic dipole moment observable.
+
+    Args:
+        mol (Molecule): the molecule object
+        cutoff (float): cutoff value for discarding the negligible dipole moment integrals
+        core (list[int]): indices of the core orbitals
+        active (list[int]): indices of the active orbitals
+
+    Returns:
+        function: function that builds the fermionic dipole moment observable
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True)
+    >>> mol = qml.hf.Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> coeffs, ops = fermionic_dipole(mol)(*args)[2]
+    >>> ops
+    [[], [0, 0], [0, 2], [1, 1], [1, 3], [2, 0], [2, 2], [3, 1], [3, 3]]
+    """
+
+    def _fermionic_dipole(*args):
+        r"""Build the fermionic dipole moment observable.
+
+        Args:
+            args (array[array[float]]): initial values of the differentiable parameters
+
+        Returns:
+            tuple(array[float], list[list[int]]): the dipole moment coefficients and the indices of
+            the spin orbitals the creation and annihilation operators act on
+        """
+        constants, integrals = dipole_integrals(mol, core, active)(*args)
+
+        nd = [anp.array([0]), anp.array([0]), anp.array([0])]
+        for i in range(len(mol.symbols)):  # nuclear contributions
+            nd[0] = nd[0] + atomic_numbers[mol.symbols[i]] * mol.coordinates[i][0]
+            nd[1] = nd[1] + atomic_numbers[mol.symbols[i]] * mol.coordinates[i][1]
+            nd[2] = nd[2] + atomic_numbers[mol.symbols[i]] * mol.coordinates[i][2]
+
+        f = []
+        for i in range(3):
+            coeffs, ops = fermionic_one(constants[i], integrals[i], cutoff=cutoff)
+            f.append((anp.concatenate((nd[i], coeffs * (-1))), [[]] + ops))
+
+        return f
+
+    return _fermionic_dipole
```

Comment:
I am a bit confused about why do we prefer building private `functions` instead of returning the values?

Reply:
It helps to separate/manipulate the differentiable parameters.